### PR TITLE
Linter for inconsistent quotation mark characters

### PR DESCRIPTION
**Pull request by megcrenshaw**
We've had two mailers hit the spam filter recently:
1) [Hour of Code signup](https://github.com/code-dot-org/code-dot-org/blob/staging/pegasus/emails/hoc_signup_2022_receipt_en.md)
2) [PD Application reminder to complete application](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/mailers/pd/application/teacher_application_mailer.rb)

The second was fixed in https://github.com/code-dot-org/code-dot-org/pull/50148. Both were/are caused by the "Right Single Quotation Mark" character (U+2019), rather than the "Apostrophe" character (U+0027), used in the subject line. See https://www.cl.cam.ac.uk/~mgk25/ucs/quotes.html for more discussion on these characters –– ASCII only supports the "Apostrophe" character and the "Quotation Mark" character.

There are three places where these errors might occur in our repo:
1) `.rb` files, which are usually configuring the subject lines 
2) `.haml` files, which usually contain the body of the message
3) `.md` files, which sometimes contain both the body and the subject line of the message.

This work assumes we don't want any right/left quotation marks in our mailers –– subject line or body. I've created custom lint rules for cases (1) and (2). For (3), I looked into this package https://www.npmjs.com/package/markdownlint-rule-search-replace. We do not have a node version compatible: `error markdownlint-rule-helpers@0.18.0: The engine "node" is incompatible with this module. Expected version ">=14.18.0". Got "14.17.1"` For now, I manually replaced these characters in all files with `*receipt*.md`

## Links

For Ruby custom rules, I used the existing examples for [string checking](https://github.com/rubocop/rubocop/blob/0699f7395e7ccebcd2ff857694b8ac96a9a13b74/lib/rubocop/cop/mixin/string_help.rb#L9). They have detailed [documentation](https://docs.rubocop.org/rubocop/1.13/development.html) too.

For Lint custom rules, I followed the [haml-lint documentation](https://github.com/sds/haml-lint#custom-linters) and used their [indentation rule](https://github.com/sds/haml-lint/blob/main/lib/haml_lint/linter/indentation.rb) as an example.

I used https://unicode-table.com/en/sets/quotation-marks/ for getting the characters to be checked, and https://www.cl.cam.ac.uk/~mgk25/ucs/quotes.html for reference.

## Testing story

Used the outputs from running the linters to check.

Ran `haml-lint` from the root directory. For haml linter running on all files, output was `873 files inspected, 114 lints detected`. For the linter running only with the `/*mailer*/` glob, `873 files inspected, 55 lints detected`.

Ran `rubocop` from the root directory. For rubocop running on all files, output was `3121 files inspected, 24 offenses detected`. For the linter running only with the `*mailer.rb` glob, 3 files threw errors. 

## Follow-up work

1) Linting for md files, possible when our node version is >=14.18.0, tracked at https://codedotorg.atlassian.net/browse/ACQ-389
2) I also found tons of mailers that are from old years, such as HOC receipts. I believe we can remove those, but the models used for these models [inherit from older models](https://github.com/code-dot-org/code-dot-org/blob/staging/pegasus/forms/hoc_signup_2017.rb#L1). It'd be great to clean this up –– tracked at https://codedotorg.atlassian.net/browse/ACQ-407

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: tools/customLinters/rubocop_consistent_quotation_mark.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+module CustomCops
+  class ConsistentQuotationMark < RuboCop::Cop::Base
+    NOT_ALLOWED_CHARACTERS = [
+      '“', # U+201C Left Double Quotation Mark
+      '”', # U+201D Right Double Quotation Mark
+      '‘', # U+2018 Left Single Quotation Mark
+      '’', # U+2019 Right Single Quotation Mark
+    ].freeze
+
+    NOT_ALLOWED_REGEX = /[#{NOT_ALLOWED_CHARACTERS.join}]/
```
Comment (megcrenshaw) [inline tools/customLinters/rubocop_consistent_quotation_mark.rb]: I tried briefly to have a constants file that both custom linters take from. I wasn't able to configure it properly –– guidance here would be great! Would I need to add something to the `application.rb` file?

**Comment by cat5inthecradle** [comment]
To clarify the problem: Is it that left/right quotes are themselves an issue, or is it that somewhere in our manipulation of the email templates the quotes are changed, causing the text of the email to not match the DKIM signature and leading spam filters to become suspicious?

**Comment by megcrenshaw** [comment]
Good question, but I'm afraid I don't have the answer.

The only two resources I've been able to dig up here are  https://powerdmarc.com/why-does-dkim-fail/ ("minor modifications in the hash value, such as the insertion of spaces or special characters can make your body hash verification fail DKIM.") and https://techcommunity.microsoft.com/t5/office-365-networking/microsoft-s-e-mailservers-break-dkim-on-non-ascii-characters/m-p/1774384 –– neither go into explanations of what exactly is causing the error.

Here's the error log of one email that hit the spam filter:
```
ARC-Authentication-Results: i=1; mx.google.com;
       dkim=fail header.i=@code.org header.s=5x6qea3fyadvw2p4ttkvlwgs4cnwvwkh header.b=VU9IFbaX;
       dkim=fail header.i=@amazonses.com header.s=6gbrjpgwjskckoa6a5zn6fwqkn67xbtw header.b=fVEkpZ6o;
       spf=pass (google.com: domain of 01000186316475a4-361d30bb-749e-4406-93f3-e5e8153c36cb-000000@amazonses.com designates 54.240.11.153 as permitted sender) smtp.mailfrom=01000186316475a4-361d30bb-749e-4406-93f3-e5e8153c36cb-000000@amazonses.com;
       dmarc=fail (p=QUARANTINE sp=QUARANTINE dis=QUARANTINE) header.from=code.org
Return-Path: <01000186316475a4-361d30bb-749e-4406-93f3-e5e8153c36cb-000000@amazonses.com>
Received: from a11-153.smtp-out.amazonses.com (a11-153.smtp-out.amazonses.com. [54.240.11.153])
        by mx.google.com with ESMTPS id b125-20020a378083000000b006fcaae20037si12648934qkd.741.2023.02.08.06.18.33
        (version=TLS1_2 cipher=ECDHE-ECDSA-AES128-GCM-SHA256 bits=128/128);
        Wed, 08 Feb 2023 06:18:33 -0800 (PST)
Received-SPF: pass (google.com: domain of 01000186316475a4-361d30bb-749e-4406-93f3-e5e8153c36cb-000000@amazonses.com designates 54.240.11.153 as permitted sender) client-ip=54.240.11.153;
Authentication-Results: mx.google.com;
       dkim=fail header.i=@code.org header.s=5x6qea3fyadvw2p4ttkvlwgs4cnwvwkh header.b=VU9IFbaX;
       dkim=fail header.i=@amazonses.com header.s=6gbrjpgwjskckoa6a5zn6fwqkn67xbtw header.b=fVEkpZ6o;
       spf=pass (google.com: domain of 01000186316475a4-361d30bb-749e-4406-93f3-e5e8153c36cb-000000@amazonses.com designates 54.240.11.153 as permitted sender) smtp.mailfrom=01000186316475a4-361d30bb-749e-4406-93f3-e5e8153c36cb-000000@amazonses.com;
       dmarc=fail (p=QUARANTINE sp=QUARANTINE dis=QUARANTINE) header.from=code.org
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/simple; s=5x6qea3fyadvw2p4ttkvlwgs4cnwvwkh; d=code.org; t=1675865913; h=To:From:Subject:List-Unsubscribe:MIME-Version:Content-Type:Message-ID:Date; bh=cayZsQqFxt+Ti5TWsdP801UX72h/cfsrDSyZ2Z44nos=; b=VU9IFbaXhnjdVovB0X02NDHZ0DiTb09fryRqf3Kl++XpfxZ2Ur0Z7PsrVIsPC2u6 P95F0HrWADOpLTwAHJnqb/cdorWweGKFKnfYHxFGsC+zghiPZlcca0dfPyjUTpm19Pd IkyQo1MBbfzJS64onmA08utMAK1E0PErk1dwklewMbGpYRS6ThIn/A2RcQG+DpTMN4X YD3GWx/wcVPo7mMeO7ARd8zYnL8Lad5vjFcq7Ix0gRrC7jmjjQ5TQXots19t9bonN4f rkooBotqR2NeqrJk460ckAUmvwxZMuS43nuCsJoUF7RiAcJ3wFTOSF2mZjqxgdb6++8 khdiK0LJcA==
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/simple; s=6gbrjpgwjskckoa6a5zn6fwqkn67xbtw; d=amazonses.com; t=1675865913; h=To:From:Subject:List-Unsubscribe:MIME-Version:Content-Type:Message-ID:Date:Feedback-ID; bh=cayZsQqFxt+Ti5TWsdP801UX72h/cfsrDSyZ2Z44nos=; b=fVEkpZ6o6EET4gGL7nuGpeHm4U+Iwxwyg+SMPGmVpWAdh3q96UesFUjmRQWG+DHj BCpKJzA+aQyNRqWQn9dcul8WP8PQ1vUuuXFC574wFfWTNcrrWfPyOXbHI3bWbRR37aY Cekc5cKg1ZkzR6eJ2K155R5+fKFSJBL7HBp5j4aw=
```

**Comment by sureshc** [comment]
> Good question, but I'm afraid I don't have the answer.

I suspect one issue is that the SMTP protocol was created before Unicode, and while Unicode and UTF-8 and non-ASCII characters are supported now in the BODY of emails through the use of MIME and `content-type-encoding` declarations on each MIME part, it’s possible that email subject’s need some special declarations to use non-ASCII characters that we’re not currently doing.

**Comment by sureshc** [comment]
Should we add a runtime check in our `Poste` module that rejects the email or logs a warning if the subject of the email contains special characters in addition to the static linter check?

**Comment by sureshc** [comment]
Looks like we can declare a character encoding for the subject line by formatting the subject [this way](https://www.telemessage.com/developer/faq/how-do-i-encode-non-ascii-characters-in-an-email-subject-line/):

`=?utf-8?Q?hello?= `

[Here’s one of the places](https://github.com/code-dot-org/code-dot-org/blob/d3a91484d5de88c2ef3a32c3fc10cc47f879c245/lib/cdo/poste.rb#L256) in `Poste` that we could modify to do this.

**Comment by sureshc** [comment]
It’s also worth thinking about building a new outbound, transactional email system just for Dashboard using [Action Mailer](https://guides.rubyonrails.org/v6.0/action_mailer_basics.html). We have to do something like this any way to support the Pegasus / Dashboard split anyway.

File: .rubocop.yml [hunk 2]
```diff
@@ -122,4 +127,4 @@ Style/SymbolArray:
 Style/ClassAndModuleChildren:
   Enabled: false
 
-# END BLACKLIST
+# END BLOCKLIST
```
Comment (davidsbailey) [inline .rubocop.yml:130]: 💯

File: tools/customLinters/rubocop_consistent_quotation_mark.rb (hunk 1 above)
Comment (davidsbailey) [reply tools/customLinters/rubocop_consistent_quotation_mark.rb]: I believe you are in plain-old-ruby land, so you shouldn't need to touch application.rb. but you could extract the shared constants into any shared ruby file which you then required into both places.

**Review by davidsbailey — COMMENTED** [review]
looks like a great approach to me! linting sounds right because we'll catch any errors as early as possible in the process, and seems compatible with the other ideas for improving our underlying email system.

**Comment by megcrenshaw** [comment]
I like these ideas. The Acq Products team is wanting to ensure we aren't using right/left quotation marks in our mailers –– but it's only hitting the spam filter with the subject lines. The Poste modification looks like it'd work if a mailer went through with a non-ASCII character.

All of our PD mailers currently inherit from ActionMailer (e.g. [TeacherApplicationMailer](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/mailers/pd/application/teacher_application_mailer.rb#L2)) –– I hope that makes it easier when the time comes to work on the Pegasus / Dashboard split with separate email systems!

File: tools/customLinters/rubocop_consistent_quotation_mark.rb (hunk 1 above)
Comment (megcrenshaw) [reply tools/customLinters/rubocop_consistent_quotation_mark.rb]: Hmm, you can see what I tried here: https://github.com/code-dot-org/code-dot-org/pull/50195/commits/fd90c131b61d0e5a0625755cf117fc0e0e26ef11.

I'm getting `uninitialized constant HamlLint::Linter::ConsistentQuotationMark::LinterConstants` and `uninitialized constant CustomCops::ConsistentQuotationMark::LinterConstants` errors. I'm seeing where we define paths to load [in our application.rb file](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/config/application.rb#L153), but it's looking like that won't apply to these files.

Comment (davidsbailey) [reply tools/customLinters/rubocop_consistent_quotation_mark.rb]: you will need to add this line before you can include the constants file:
```ruby
require_relative 'linter_constants'
```

application.rb is part of rails. my assumption is that the linter runs outside of rails, so it doesn't care what's in application.rb. instead, you have to use pure ruby to access the file.

in rails, everything within the rails directories is `require`d for you. outside of rails (or in rails and when accessing things that are outside of rails) you must `require` a file before you can access it. `include` is not a substitute for `require`, in fact once you are requiring, you could drop the `include` and just use `LinterConstants::NOT_ALLOWED_CHARACTERS` to access your constant.

clear as mud?

Comment (megcrenshaw) [reply tools/customLinters/rubocop_consistent_quotation_mark.rb]: Wow, that was very helpful, thank you! Got it working, tested on both "bad" ruby and haml files.

File: tools/customLinters/linter_constants.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  NOT_ALLOWED_CHARACTERS = [
+    '“', # U+201C Left Double Quotation Mark
+    '”', # U+201D Right Double Quotation Mark
+    '‘', # U+2018 Left Single Quotation Mark
+    '’', # U+2019 Right Single Quotation Mark
+  ].freeze
+
+  NOT_ALLOWED_REGEX = /[#{NOT_ALLOWED_CHARACTERS.join}]/
+
+  MSG = "Do not use left/right quotation mark. Use \' or \" instead.".freeze
```
Comment (davidsbailey) [inline tools/customLinters/linter_constants.rb]: nit: the constant names `MSG` and others seem a bit too general for something living in a module named `LinterConstants`, and if someone adds something to NOT_ALLOWED_CHARACTERS that isn't a quote, they'll get a confusing error message when they hit the new character.

what do you think about changing this to include the actual list of disallowed characters in the error message? that'll let people add to the list without any surprises.

if you're not set on reusability (which is fine) a couple of alternatives would be to (1) rename the module to include the word "quote" or (2) rename each of these constants to include the word "quote".

**Review by davidsbailey — APPROVED** [review]
very nice! one more nit then looks good :-D

File: tools/customLinters/linter_constants.rb (hunk 3 above)
Comment (megcrenshaw) [reply tools/customLinters/linter_constants.rb]: Good call. `MSG` is needed for Rubocop custom linters, but I changed the constant in the file to be `NOT_ALLOWED_MSG` and then pointed `MSG` to it in the Rubocop linter. I changed the file names and error message string to be generic for any not-allowed characters. Now the error messages look like

in haml-lint:
```
dashboard/app/views/pd/application/teacher_application_mailer/accepted.html.haml:9 [W] OnlyAllowedCharacters: The following characters are not allowed: “ ” ‘ ’
```

and in Rubocop:
```
dashboard/app/mailers/pd/application/teacher_application_mailer.rb:139:18: C: CustomCops/OnlyAllowedCharacters: The following characters are not allowed: “ ” ‘ ’
        subject: "Reminder: Complete your application for Code.org‘s Professional Learning Program"
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Comment (megcrenshaw) [reply tools/customLinters/linter_constants.rb]: Open to other refactors! Thanks

File: tools/customLinters/rubocop_only_allowed_characters.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+require_relative 'linter_constants'
+
+module CustomCops
+  class OnlyAllowedCharacters < RuboCop::Cop::Base
+    include LinterConstants
+
+    MSG = NOT_ALLOWED_MSG
```
Comment (davidsbailey) [inline tools/customLinters/rubocop_only_allowed_characters.rb:7]: wow, I had no idea the Cop was depending on `MSG` being defined! it looks better to define it explicitly like this

File: tools/customLinters/linter_constants.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  NOT_ALLOWED_CHARACTERS = [
+    '“', # U+201C Left Double Quotation Mark
+    '”', # U+201D Right Double Quotation Mark
+    '‘', # U+2018 Left Single Quotation Mark
+    '’', # U+2019 Right Single Quotation Mark
+  ].freeze
+
+  NOT_ALLOWED_REGEX = /[#{NOT_ALLOWED_CHARACTERS.join}]/
+
+  NOT_ALLOWED_MSG = "The following characters are not allowed: #{NOT_ALLOWED_CHARACTERS.join(' ')}".freeze
```
Comment (davidsbailey) [inline tools/customLinters/linter_constants.rb:11]: perfect!

**Review by davidsbailey — APPROVED** [review]
updates look great 👍

**Pull request merged**